### PR TITLE
BUG: groupby.shift returns different columns when fill_value is specified

### LINKED ISSUE #26615: Performance issue while groupby.shift if fill_value specified
Test code:
```python
    SIZE_MULT = 5
    data = np.random.randint(0, 255, size=10**SIZE_MULT, dtype='uint8')
    index = pd.MultiIndex.from_product(
                [list(range(10**(SIZE_MULT-1))), list('ABCDEFGHIJ')],
                names = ['d', 'l'])        
    test = pd.DataFrame(data, index, columns = ['data'])
    test.head()
    test['data'].dtype
```
```python
     		data
    d 	l 	
    0 	A 	137
        B 	156
        C 	48
        D 	186
        E 	170
    
    dtype('uint8')
```
And suppose we want group by 0-level of index and shift each group (shift step = 2, for example). 
```python
%%time
shifted = test.groupby(axis=0, level=[0]).shift(2)
print(shifted['data'].dtype)
```
```
float64
CPU times: user 9.43 ms, sys: 56 µs, total: 9.49 ms
Wall time: 8.29 ms
```
Now to the problem: if we want to preserve our dtype 'uint8', we have to get rid of `None`s, and set our fill value with 0, for example. But we will get HUGE time of code execution now:
```python
%%time
shifted = test.groupby(axis=0, level=[0]).shift(2, fill_value = 0)
shifted.head()
print(shifted['data'].dtype)
```
```
uint8
CPU times: user 5.9 s, sys: 38.4 ms, total: 5.94 s
Wall time: 5.89 s
```
Expected time of execution should be comparable with following:
 If we take 1st shifted dataframe without `fill_value`, and add few code lines to achieve same result:
```python
%%time
shifted = test.groupby(axis=0, level=[0]).shift(2)
shifted.fillna(0, inplace=True)
shifted = shifted.astype(np.uint8)
print(shifted['data'].dtype)
```
**Output:**
```
uint8
CPU times: user 9.64 ms, sys: 3.68 ms, total: 13.3 ms
Wall time: 11.3 ms
```
It will add only few ms, not 5 seconds.

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Linux
OS-release: 5.0.0-16-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 4.5.0
pip: 19.1.1
setuptools: 41.0.1
Cython: 0.29.7
numpy: 1.16.3
scipy: 1.2.1
pyarrow: None
xarray: 0.12.1
IPython: 7.2.0
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: 3.5.1
numexpr: 2.6.9
feather: None
matplotlib: 3.0.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10.1
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### LINKED ISSUE #41556: BUG: groupby.shift return keys filled with `fill_value` when `fill_value` is specified
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
In [2]: df = pd.DataFrame({'a': [2, 1, 2, 1], 'b': ['x', 'x', 'y', 'y']})

In [3]: df.groupby('a').shift(1)
Out[3]: 
     b
0  NaN
1  NaN
2    x
3    x

In [4]: df.groupby('a').shift(1, fill_value='fill')
Out[4]: 
      a     b
0  fill  fill
1  fill  fill
2     2     x
3     1     x
```

#### Problem description
When specifying `fill_value` in `groupby.shift`, the returned result includes the key column with keys filled with `fill_value`. When `fill_value` is unspecified (None), the key column is not included.

#### Expected Output
It seems pretty strange that keys are to be filled with `fill_value`. This makes more sense to me:
```python
In [4]: df.groupby('a').shift(1, fill_value='fill')
Out[4]: 
      b
0  fill
1  fill
2    x
3    x
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.7.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-76-generic
Version          : #86-Ubuntu SMP Fri Jan 17 17:24:28 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8

pandas           : 1.2.4
numpy            : 1.20.2
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1.1
setuptools       : 52.0.0.post20210125
Cython           : 0.29.23
pytest           : 6.2.4
hypothesis       : 6.12.0
sphinx           : 3.5.4
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 7.23.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 2021.04.0
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 1.0.1
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.53.1
</details>

### PR DESCRIPTION
**Pull request by smithto1**
- [x] closes #41556
- [x] closes #26615 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

This fixes the minimal reproducing examples from the original bug report #41556. 

**With master**, specifying `fill_value` causes the index columns to be returned and to include the `fill_value` in the grouping columns. 

```
In [4]: import pandas as pd
   ...: 
   ...: df = pd.DataFrame({'a': [2, 1, 2, 1], 'b': ['x', 'x', 'y', 'y']})
   ...: 
   ...: df.groupby('a').shift(1)
Out[4]:
     b
0  NaN
1  NaN
2    x
3    x
In [5]: df.groupby('a').shift(1, fill_value='fill')
Out[5]:
      a     b
0  fill  fill
1  fill  fill
2     2     x
3     1     x
```

**With this PR**, only the value columns are returned in both cases, and the fill is applied correctly.

```
In [1]: import pandas as pd
   ...: 
   ...: # On master, using fill_value causes the index columns to be returned
   ...: df = pd.DataFrame({'a': [2, 1, 2, 1], 'b': ['x', 'x', 'y', 'y']})
   ...: 
   ...: df.groupby('a').shift(1)
Out[1]:
     b
0  NaN
1  NaN
2    x
3    x
In [2]: df.groupby('a').shift(1, fill_value='fill')
Out[2]:
      b
0  fill
1  fill
2     x
3     x
```

On master, if `fill_value=None` then `_get_cythonized_result` was used. But if `fill_value` was specified, then `self.apply` was used, because `_get_cythonized_result` couldn't take the `fill_value`. I've updated `_get_cythonized_result` so it can handle the `fill_value` itself. This means `Groupby.shift` follows the same code path and returns the same structure of output, whether or not `fill_value` is specified. 

**Test**
The bug reported in the original issue should have been caught by an existing test, except that the existing test extracts the values columns ( `[['Z']]` ), before doing its comparison to the expected output. This was hiding the fact that the structure of the result depended on `fill_value` being specified. 

https://github.com/pandas-dev/pandas/blob/1.2.x/pandas/tests/groupby/test_groupby_shift_diff.py#L52

Removing the extraction of the data column ( `[['Z']]` ) means we now have a test that catches the bug reported in the original issue. (The test fails on master, but passes on this PR).

### CODE REVIEW

**Comment by WillAyd** [comment]
Thanks for the PR. I _think_ that dispatching to `self.apply` here is going to cause a rather large performance regression though - can check the benchmarks to see?

We likely need to patch elsewhere in the groupby code to make ensure that the correct structure is returned

**Comment by smithto1** [comment]
> Thanks for the PR. I _think_ that dispatching to `self.apply` here is going to cause a rather large performance regression though - can check the benchmarks to see?
> 
> We likely need to patch elsewhere in the groupby code to make ensure that the correct structure is returned

I think there is a misunderstanding here. I have removed the dispatch to `self.apply` and switched it to `_get_cynthonized_result`. I haven't checked but would assume performance would improve.

**Comment by simonjayhawkins** [comment]
> @WillAyd, I think there is a misunderstanding here. I have removed the dispatch to `self.apply` and switched it to `_get_cynthonized_result`. I haven't checked but would assume performance would improve.

does this close #26615

**Review by WillAyd — APPROVED** [review]
Thanks for clarifying. This lgtm

**Comment by smithto1** [comment]
@WillAyd Is there anything else we need to have this merged in?

**Review by jreback — CHANGES_REQUESTED** [review]
needs tests that reproduce the original issues
release note needs to be update to 1.4
needs review

**Comment by smithto1** [comment]
> needs tests that reproduce the original issues

This PR includes a modification to an existing test which ensures it recreates the original issue. The modified test fails on master (see below and attached) because of the original issue. With the code changes in this PR the test now passes. 

Previously, the test extracted the value column ( `[['Z']]` ) from the `result`, ignoring the fact that the `result` contained additional columns that should have caused an error. This is fixed now. 

```
[err1429.txt](https://github.com/pandas-dev/pandas/files/6648622/err1429.txt)

        result = g.shift(-1, fill_value=0)
    
>       tm.assert_frame_equal(result, expected)
E       AssertionError: DataFrame are different
E       
E       DataFrame shape mismatch
E       [left]:  (24, 3)
E       [right]: (24, 1)

test_groupby_shift_diff.py:60: AssertionError
```

> release note needs to be update to 1.4

Done. 

> needs review

Ready for review, @jreback

**Comment by smithto1** [comment]
@jreback can you take another look at this? I think it is ready to merge.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. can you make sure we have an asv for this case (or add one) and show the results. ping on green. 

I think testing is ok, though if you can look in the original issue and make sure we have a replicated test.

**Comment by smithto1** [comment]
> looks good. can you make sure we have an asv for this case (or add one) and show the results. ping on green.
> 
> I think testing is ok, though if you can look in the original issue and make sure we have a replicated test.

There is a replicating test. 

asv is added. Can see the difference running the same asv for master and for the PR branch. 

**master**
```
[  0.00%] ·· Benchmarking existing-pyc__anaconda3_envs_pandas-dev_python.exe
[ 25.00%] ··· groupby.Shift.time_defaults                              1.89±0ms
[ 50.00%] ··· groupby.Shift.time_fill_value                            3.47±0ms
```

**PR**
```
[  0.00%] ·· Benchmarking existing-pyc__anaconda3_envs_pandas-dev_python.exe
[ 25.00%] ··· groupby.Shift.time_defaults                              1.33±0ms
[ 50.00%] ··· groupby.Shift.time_fill_value                            1.29±0ms
```

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -187,7 +187,7 @@ Plotting
 
 Groupby/resample/rolling
 ^^^^^^^^^^^^^^^^^^^^^^^^
--
+- Bug in :meth:`GroupBy.shift` that would return different columns if ``fill_value`` was not None (:issue:`41556`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.4.0.rst:190]: Can be a little more specific, different columns -> the group by columns

Comment (rhshadrach) [reply doc/source/whatsnew/v1.4.0.rst:190]: Also - can you add a line in the performance section.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
lgtm, minor request for the whatsnew

**Comment by rhshadrach** [comment]
@smithto1 - friendly ping for resolving conflict and whatsnew request.

**Comment by smithto1** [comment]
> @smithto1 - friendly ping for resolving conflict and whatsnew request.

That is addressed. I think this is ready to merge. 

@rhshadrach @jreback @WillAyd

**Review by rhshadrach — APPROVED** [review]
lgtm

File: doc/source/whatsnew/v1.4.0.rst (hunk 1 above)
Comment (smithto1) [reply doc/source/whatsnew/v1.4.0.rst]: Done

File: pandas/tests/groupby/test_groupby_shift_diff.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def test_group_shift_with_fill_value():
         columns=["Z"],
         index=None,
     )
-    result = g.shift(-1, fill_value=0)[["Z"]]
+    result = g.shift(-1, fill_value=0)
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby_shift_diff.py:58]: do we have sufficient coverage of the tests from the OP?

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. test coverage question. ping on green.

File: pandas/tests/groupby/test_groupby_shift_diff.py (hunk 2 above)
Comment (smithto1) [reply pandas/tests/groupby/test_groupby_shift_diff.py:58]: Yes. I have investigated it. 

The original problem was that the grouping columns were improperly returned. The old form of this test would extract the value column (`[["Z"]]`), so the test did not detect that the grouping columns were also returned. I have modified this test so it no longer extracts the value column; if the grouping columns are returned (the OP) this test will now fail.

Comment (jreback) [reply pandas/tests/groupby/test_groupby_shift_diff.py:58]: kk great

if u can resolve conflicts and merge master

Comment (smithto1) [reply pandas/tests/groupby/test_groupby_shift_diff.py:58]: @jreback conflicts resolved. I think it can be merged.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @smithto1 for the patch and the patience! keep em coming!

**Pull request merged**